### PR TITLE
rewrote dataset and nn classes

### PR DESCRIPTION
**Pull request by yuanqing-wang**

### CODE REVIEW

**Comment by yuanqing-wang** [comment]
I do not to intent to merge this into master soon. But something like this would enable us to do training in grammar like:

https://github.com/choderalab/espaloma/blob/unify/scripts/typing/playground.py

We can of course further abstract this.

**Comment by yuanqing-wang** [comment]
Specifically, we can further,

-[] abstract metrics / loss to enable seamless training and testing
-[] abstract experiment into objects

**Comment by yuanqing-wang** [comment]
I'm also thinking about carrying over some CLI defined here:

https://github.com/choderalab/pinot/tree/master/pinot/app

But this would be inherently harder as now we are facing multiple metrics.

**Comment by yuanqing-wang** [comment]
I will merge this today at 5 pm EST if there's no objection. @maxentile

File: espaloma/data/dataset.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        objects in the dataset
+
+    Note
+    ----
+    This also supports iterative-style dataset by deleting `__getitem__`
+    and `__len__` function. 
+
+    Attributes
+    ----------
+    transforms : an iterable of callables that transforms the input.
```
Comment (maxentile) [inline espaloma/data/dataset.py:26]: Add a bit more description or example here -- for example, the `__getitem__` definition applies these transforms sequentially, so order may matter

File: espaloma/data/dataset.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+    """
+    def __init__(self, graphs=None):
+        super(Dataset, self).__init__()
+        self.graphs = graphs
+        self.transforms = None
```
Comment (maxentile) [inline espaloma/data/dataset.py:32]: Where do `self.transforms` get assigned, since not added in constructor?

File: espaloma/data/dataset.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            # is this efficient?
+            graphs = iter(self.graphs)
+            for transform in self.transforms:
+                graphs = map(
+                        transform,
+                        graphs)
+
+            return graphs
+
+    def apply(self, fn, in_place=False):
```
Comment (maxentile) [inline espaloma/data/dataset.py:89]: Docstring would be helpful does `in_place` mean the graphs are modified in place, or that this modifies self.transforms, or both?

File: espaloma/data/dataset.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+                self.transforms = []
+
+            self.transforms.append(fn)
+
+        else: # modify in-place
+            self.graphs = list(map(fn, self.graphs))
+
+        return self # to allow grammar: ds = ds.apply(...)
+
+    def split(self, partition):
```
Comment (maxentile) [inline espaloma/data/dataset.py:104]: docstring clarifying what `partition` is expected to be (sequence of floats, not required to sum to one, I think?). can also comment on why pytorch data splitting utilities like https://pytorch.org/docs/stable/data.html#torch.utils.data.random_split are unsuitable

File: espaloma/data/dataset.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+                    graph, 
+                    Molecule
+                ) or isinstance(
+                    graph,
+                    str) for graph in graphs):
+
+            if first is None:
+                graphs = [esp.Graph(graph) for graph in graphs]
+            else:
+                graphs = [esp.Graph(graph) for graph in graphs[:first]]
```
Comment (maxentile) [inline espaloma/data/dataset.py:140]: document intent of `first` argument and `[:first]` slice -- for example, why not call constructor with different input graphs, `GraphDataset(graphs)` vs `GraphDataset(graphs[:first])`

File: espaloma/data/dataset.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+            return dgl.batch(self.graphs)
+        
+        elif all(isinstance(graph, dgl.DGLHeteroGraph) for graph in graphs):
+            return dgl.batch_hetero(self.graphs)
+
+        else:
+            raise RuntimeError('Can only batch DGLGraph or DGLHeterograph,'
+                'now have %s' % type(graphs[0]))
+
+    def view(self, collate_fn='graph', *args, **kwargs):
```
Comment (maxentile) [inline espaloma/data/dataset.py:160]: docstring

File: espaloma/data/dataset.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+                y = graph.ndata['legacy_typing']
+                return graph, y
+
+        elif collate_fn == 'graph-typing-loss':
+            loss_fn = torch.nn.CrossEntropyLoss()
+            def collate_fn(graphs):
+                graph = self.batch(graphs)
+                loss = lambda _graph: loss_fn(
+                        _graph.ndata['nn_typing'],
+                        graph.ndata['legacy_typing'])
```
Comment (maxentile) [inline espaloma/data/dataset.py:176]: "view of dataset" also requires access to private attribute `_graph.ndata['nn_typing']` containing model prediction

File: scripts/typing/playground.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+def run():
+    # grab dataset
+    esol = esp.data.esol(first=20)
+
+    # do some typing
+    typing = esp.graphs.legacy_force_field.LegacyForceField('gaff-1.81')
+    esol.apply(typing, in_place=True) # this modify the original data
+
+    # split
```
Comment (maxentile) [inline scripts/typing/playground.py:12]: can move this comment into .split definition

File: espaloma/data/dataset.py (hunk 5 above)
Comment (yuanqing-wang) [reply espaloma/data/dataset.py:140]: ah yeah this looks a bit redundant indeed

Comment (yuanqing-wang) [reply espaloma/data/dataset.py:140]: okay I think I was thinking about adding it to the pre-written collection of datasets

**Comment by yuanqing-wang** [comment]
fixed these. thanks!
